core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -388,7 +388,13 @@ private[spark] class MemoryStore(
     // perform one final call to attempt to allocate additional memory if necessary.
     if (keepUnrolling) {
       serializationStream.close()
-      reserveAdditionalMemoryIfNecessary()
+      if (bbos.size > unrollMemoryUsedByThisBlock) {
+        val amountToRequest = bbos.size - unrollMemoryUsedByThisBlock
```

Comment:
Why not just fix that in `reserveAdditionalMemoryIfNecessary` ?

Reply:
Because we request `val amountToRequest = (bbos.size * memoryGrowthFactor - unrollMemoryUsedByThisBlock).toLong` in `reserveAdditionalMemoryIfNecessary ` to avoid requesting for every records.  But here we just need request the precise memory for the last request.